File: src/CRF/python/crf_tag.py
#!/usr/bin/env python
"""
author : Sonali Rahagude (srahagud@eng.ucsd.edu)
Script to train CRF, accepts input file which is complaint to CRFSuite requirements
"""

import crfsuite
import sys
import pandas as pd
import crf_train
import util

def write_prediction_to_file (prediction_seq, label_seq, overlapped_predictions, output, options_dict):

    crf_sequence_length = options_dict ['crf_sequence_length']
    sliding_window_length = options_dict ['sliding_window_length']

    '''
    if len(overlapped_predictions) == 0 :
        if crf_sequence_length != len(prediction_seq):
            # write from 0 to len(prediction_seq)
            for i in range(0: legnt)
            overlapped_predictions = []
            return overlapped_predictions
        else:         
            # write from 0 to (prediction_seq - sliding window length)
            new_start = crf_sequence_length - sliding_window_length
            overlapped_predictions = prediction_seq[new_start:len(prediction_seq)]
            return overlapped_predictions
    '''
    # voting mechanism: we can find overlap between subsequent sequences as follows,
    ## 1. for the overlap received from the previous sequence, voting can be done and written
    ## 2. for no overlap region, the prediction can be written as are, to the output file
    ## 2. if the sequence is not cut off, it will have some overlap with the next sequence. So the overlap can be passed to the next
    ##    of this method for voting

    # I assume that the length of the prediction sequence should be at least the length of the overlapped predictions. This will 
    # fail if the sequence cut off exactly at the provided sequence length (crf_sequence_file). 
    # Need to have an exact way of finding overlaps. Sequences could have an additional feature in the input file
    # telling if shares an overlap with the next one.
    for i in range(0, len(overlapped_predictions)) :

        if prediction_seq[i] != overlapped_predictions[i] :
            
            print ("applying voting. i: ",i,", prediction_seq[i]: ", prediction_seq[i], ", overlapped_predictions[i]: ",overlapped_predictions[i],"label_seq[i]: ",label_seq[i],"\n")
            similarity_cur = util.get_similarity(prediction_seq[i],prediction_seq[0:len(prediction_seq)])
            similarity_prev = util.get_similarity(overlapped_predictions[i], overlapped_predictions)
            print("similarity_cur: ", similarity_cur, ", similarity_prev: ", similarity_prev, "\n")
            
            if similarity_prev < similarity_cur :
                output.write( prediction_seq[i] + "," + label_seq[i] +"\n")
            else :
                output.write( overlapped_predictions[i] + "," + label_seq[i] +"\n")

        else :
            output.write( prediction_seq[i] + "," + label_seq[i] +"\n")

    no_overlap_region = min(crf_sequence_length - sliding_window_length, len(prediction_seq))

    for i in range(len(overlapped_predictions), no_overlap_region) :
        output.write( prediction_seq[i] + "," + label_seq[i] +"\n")

    if(no_overlap_region == len(prediction_seq)):
        overlapped_predictions = []
        return overlapped_predictions

    overlapped_predictions = prediction_seq[no_overlap_region:crf_sequence_length]
    return overlapped_predictions
   
   
'''
    for i in range(0, actual_overlap) :
        if len(overlapped_predictions) == 0 :
                output.write( prediction_seq[i] + "," + label_seq[i] +"\n")
        else :
            # apply voting
            if prediction_seq[i] != overlapped_predictions[i] :
                
                print ("applying voting. i: ",i,", prediction_seq[i]: ", prediction_seq[i], ", overlapped_predictions[i]: ",overlapped_predictions[i],"label_seq[i]: ",label_seq[i],"\n")
                similarity_cur = util.get_similarity(prediction_seq[i],prediction_seq[0:actual_overlap])
                similarity_prev = util.get_similarity(overlapped_predictions[i], overlapped_predictions)
                print("similarity_cur: ", similarity_cur, ", similarity_prev: ", similarity_prev, "\n")
                
                if similarity_prev < similarity_cur :
                    output.write( prediction_seq[i] + "," + label_seq[i] +"\n")
                else :
                    output.write( overlapped_predictions[i] + "," + label_seq[i] +"\n")

            else :
                output.write( prediction_seq[i] + "," + label_seq[i] +"\n")

    if len(prediction_seq) != crf_sequence_length : 
        overlapped_predictions = []
        return overlapped_predictions

    new_start = crf_sequence_length - sliding_window_length
    overlapped_predictions = prediction_seq[new_start:len(prediction_seq)]
    return overlapped_predictions
'''
'''
def read_file_to_crfsuite(crf_input_file, feature_inclusion_list, crf_tagger, output, options_dict):    
    import crfsuite
    f = open(crf_input_file, 'r')
    min_max  = pd.load('min_max_dataframe')
    xseq = crfsuite.ItemSequence()
    yseq = crfsuite.StringList()
    for line in f:        
        if "label" in line:            
            feature_index_list = crf_train.get_feature_index_list(line, feature_inclusion_list)
            header = line.split('\t')
            continue
        if "START" in line:
            continue
        if "END" in line:
            #crf_tagger.set(xseq)    
            y_itr = yseq.iterator()
            for prediction in crf_tagger.tag(xseq):
                label = y_itr.next()
                #print "prediction: " + prediction
                #print "label: " + label
                output.write(prediction.strip() + "," + label.strip()+"\n")
                #print prediction.strip() + "," + label.strip()
            xseq = crfsuite.ItemSequence()
            yseq = crfsuite.StringList()
        else:
            item = crfsuite.Item()
            fields = line.split('\t')
            for i in range(0,len(fields)):
                if i in feature_index_list:
                    # print header[i]
                    # print fields[i]
                    attribute_name = header[i]
                    if(fields[i] == 'NA'):
                        attribute_val = 0
                    else:
                        #attribute_val = float(fields[i]) 
                        denom = min_max['max'][attribute_name] - min_max['min'][attribute_name]
                        if denom == 0:
                            attribute_val = 0
                        else: 
                            attribute_val = (float(fields[i]) - min_max['min'][attribute_name] )/ denom
                        #print attribute_val
                    item.append(crfsuite.Attribute(attribute_name, attribute_val))            
            xseq.append(item)
            #print xseq.items()
            yseq.append(fields[0])

               

'''
def read_file_to_crfsuite(crf_input_file, feature_inclusion_list, crf_tagger, output, options_dict):    
    sliding_window_length = options_dict['sliding_window_length']
    import crfsuite
    f = open(crf_input_file, 'r')
    #min_max  = pd.load('min_max_dataframe')
    xseq = crfsuite.ItemSequence()
    yseq = crfsuite.StringList()
    overlapped_predictions = []
    for line in f:        
        if line.strip(' \t\n\r')=="":
            continue
        if "label" in line:
            feature_index_list = crf_train.get_feature_index_list(line, feature_inclusion_list)
            header = line.split('\t')
            continue  
        if "START" in line:
            continue        
        if "END" in line:
            #crf_tagger.set(xseq)    
            #prediction_seq = util.convert_to_python_list(crf_tagger.tag(xseq))
            prediction_seq = crf_tagger.tag(xseq)
            label_seq = util.convert_to_python_list (yseq)
            if (sliding_window_length != 0 ):
                overlapped_predictions = write_prediction_to_file (prediction_seq, label_seq, overlapped_predictions, output , options_dict)
            else:
                y_itr = yseq.iterator()
                for prediction in prediction_seq:
                    #print 'straightforward'
                    label = y_itr.next()   
                    output.write(prediction.strip() + "," + label.strip()+"\n")

            xseq = crfsuite.ItemSequence()
            yseq = crfsuite.StringList()
        else:
            item = crfsuite.Item()
            fields = line.split('\t')
            for i in range(0,len(fields)):
                if i in feature_index_list:
                    attribute_name = header[i]
                    if(fields[i] == 'NA'):
                        attribute_val = 0
                    else:
                        attribute_val = float(fields[i])                    
                    item.append(crfsuite.Attribute(attribute_name, attribute_val))
            xseq.append(item)            
            yseq.append(fields[0].strip('"'))

def crf_tag(crf_model_file, crf_test_file, feature_list_file, output_file, options_dict):

    output = open(output_file, 'w')
	# Create a tagger object.
    tagger = crfsuite.Tagger()
    
    # Load the model to the tagger.
    tagger.open(crf_model_file)

    output.write("Prediction,True Label\n")
    feature_inclusion_list = crf_train.get_feature_inclusion_list(feature_list_file)

    read_file_to_crfsuite(crf_test_file,feature_inclusion_list, tagger,output, options_dict)
    """
    for xseq in instances(fi):
    	# Tag the sequence.
        tagger.set(xseq)
        # Obtain the label sequence predicted by the tagger.
        yseq = tagger.viterbi()
        # Output the probability of the predicted label sequence.
        print tagger.probability(yseq)
        for t, y in enumerate(yseq):
        	# Output the predicted labels with their marginal probabilities.
            print '%s:%f' % (y, tagger.marginal(y, t))
        print
    """

File: src/CRF/R/crf_tag.R
# author : Sonali Rahagude (srahagud@eng.ucsd.edu)
# Description: This module implements prediction part of the crf

# Predicts sequences for test data based on the CRF model passed to it
# Arguments:
## crf_test_file: file containing the crf sequences along with the labels. Each line is of the form <label, feature1, feature2,...featuren>
##                 each sequence is enclosed between tags #START and #END
## crf_model_file: filename to read the weights of feature functions from
## feature_list_file: contains list of names of feature functions to be used to generate features
## output_prediction_file: file to save the predictions and actual labels
## options: A dictionary contains values for miscellaneous parameters

crf_tag = function(crf_test_file, crf_model_file, feature_list_file, output_prediction_file, options, test_list = NULL) {

    model = list()
	# This will load weights_new object from the saved training model
	load(crf_model_file)
    crf_weights = model$weights
    pre_proc_values = model$pre_proc_values
    #print(crf_weights)
	feature_inclusion_list = get_feature_inclusion_list(feature_list_file)
    #print(feature_inclusion_list)

	# read training data from file
    if(is.null(test_list)) {        
        #cat("Reading from file: ", crf_test_file,"\n")
	   test_list = get_training_data(crf_test_file, feature_inclusion_list)
    } 
	x_list = test_list[[1]]
	y_list = test_list[[2]]

    pre_proc_values = get_pre_proc_values(x_list)
    x_list = get_scaled_training_data(x_list, pre_proc_values)

	# get the template features. These will be combined with indicator functions I(y[i-1]='label1') * I(y[i]='label2') 
	features = colnames(x_list[[1]])
	# get the list of labels from the dimension names of the weights array
	labels = dimnames(crf_weights)[[2]]
    
    # just in case the feature list from training does not conform
     crf_weights = crf_weights[,,unlist(features)]
	
    predictions = file.path(output_prediction_file)
	if (file.exists( predictions) ) {
      file.remove(output_prediction_file)
    }
    
    crf_sequence_length = as.numeric(options["crf_sequence_length"])
    overlap_window_length = as.numeric(options["overlap_window_length"])    
    overlapped_predictions = list()
    overlapped_labels = list()
	
    cat('prediction,label\n', file=predictions, append=TRUE)	
    ii = 0
    for(j in 1:length(x_list)) {
        x = x_list[[j]]
        y = y_list[[j]]
        if (length(y) != nrow(x)) {
            cat('Error with Preprocessing, lengths x and y dont match\n');          
            return
        }
        #cat("taggging example: ", j, "\n")
        #Calculate Viterbi Path and y_hat, uses anotherCOMPUTEG               
        G = compute_g_matrices(x, crf_weights, labels)
        #print(G)
 		yHat = viterbi(G, x, labels) 
        # print(x)
        # print (yHat)
        # print("actual y")
        # print (y)    
        # --------------------------------------------------------------------
        for(i in 1:length(yHat)) {
            cat(yHat[i], y[i], sep=",", file=predictions, append=TRUE)
            cat('\n', file=predictions, append = TRUE)
        }
        # --------------------------------------------------------------------
        
        # --------------------------------------------------------------------
        # voting mechanism
        # actual_overlap = min(overlap_window_length, length(yHat))
        # print(actual_overlap)
        # for(i in 1: actual_overlap) {
        #     if(length(overlapped_predictions) == 0) {
        #             cat(yHat[i], y[i], sep=",", file=predictions, append=TRUE)
        #             cat('\n', file=predictions, append = TRUE)
        #     }
        #     else {
        #         # apply voting
        #         if(yHat[i] != overlapped_predictions[i]) {
        #             ii = ii + 1
        #             cat("applying voting. i: ",i,", yHat[i]: ", yHat[i], ", overlapped_predictions[i]: ",overlapped_predictions[i],"y[i]: ",y[i],"\n")
        #             similarity_cur = get_similarity(yHat[i],yHat[1:actual_overlap])
        #             similarity_prev = get_similarity(overlapped_predictions[i], overlapped_predictions)
        #             cat("similarity_cur: ", similarity_cur, ", similarity_prev: ", similarity_prev, "\n")
        #             print(overlapped_predictions)
        #             print(yHat[1:actual_overlap])
        #             print(y[1:actual_overlap])
        #             if(similarity_prev < similarity_cur) {
        #                 cat(yHat[i], y[i], sep=",", file=predictions, append=TRUE)
        #                 cat('\n', file=predictions, append = TRUE)
        #             } else {
        #                 cat(overlapped_predictions[i], y[i], sep=",", file=predictions, append=TRUE)
        #                 cat('\n', file=predictions, append = TRUE)
        #             }
        #         }
        #         else {
        #             cat(yHat[i], y[i], sep=",", file=predictions, append=TRUE)
        #             cat('\n', file=predictions, append = TRUE)
        #         }
        #     }
        # }
        # if(length(yHat) <= overlap_window_length) {
        #     next
        # }
        # new_start = overlap_window_length + 1
        # overlapped_predictions = yHat[new_start:length(yHat)]
        # overlapped_labels = y[new_start:length(y)]
        # if (length(yHat) != crf_sequence_length) {
        #     overlapped_predictions = list()
        #     overlapped_labels = list()
        # }
        # --------------------------------------------------------------------
    }    
    cat("total conflicts: ",ii,"\n")
}


get_similarity = function(label, label_sequence) {
    sim = sum(label_sequence == label)
}

